### PR TITLE
[TEP-0089] SPIRE for non-falsifiable provenance.

### PR DESCRIPTION
**Pull request by jagathprakash**
[TEP-0089] SPIRE for non-falsifiable provenance.

This PR is a part of a larger set of PRs to provide non-falsifiable provenance through SPIRE.
In particular this PR uses the SPIRE infrastructure which has already been merged to sign TaskRunStatus.
It also has support to verify if TaskRunStatus has been modified by another workload between reconciles.

Previously merged PRs are
- [PR-5039](https://github.com/tektoncd/pipeline/pull/5039)
- [PR-5647](https://github.com/tektoncd/pipeline/pull/5647)
- [PR-5676](https://github.com/tektoncd/pipeline/pull/5676)
- [PR-5902](https://github.com/tektoncd/pipeline/pull/5902)

# Changes

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [X] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [X] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [X] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [X] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [X] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [X] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [X] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

- Added TaskRun Status annotations to track the validity of the signed TaskRun
- Utilizes pipeline controller spire SVID, status hash and signature
- Pipeline controller continuously validates the TaskRun Status for any modifications
- Tekton Chains will validate the results and status of the TaskRun after completion

### CODE REVIEW

**Comment by jagathprakash** [comment]
@lumjjb @pxp928 This is a merge of phase1 and phase2. This has a few changes to code as well as unit tests on top of the phase1 and phase2 PRs. Please take a look.

File: cmd/imagedigestexporter/main.go [hunk 1]
```diff
@@ -31,6 +34,8 @@ import (
 var (
 	images                 = flag.String("images", "", "List of images resources built by task in json format")
 	terminationMessagePath = flag.String("terminationMessagePath", "/tekton/termination", "Location of file containing termination message")
+	enableSpire            = flag.Bool("enable_spire", false, "If specified by configmap, this enables spire signing and verification")
```
Comment (jerop) [inline cmd/imagedigestexporter/main.go:37]: do we still use `enable-spire` feature flag? thought we changed it to `enforce-nonfalsifiability` in https://github.com/tektoncd/pipeline/pull/5902 in line with the TEP?

File: examples/v1beta1/pipelineruns/4808-regression.yaml [hunk 2]
```diff
@@ -92,4 +92,4 @@ spec:
     name: result-test
   params:
   - name: RESULT_STRING_LENGTH
-    value: "3000"
+    value: "2000"
```
Comment (jerop) [inline examples/v1beta1/pipelineruns/4808-regression.yaml:95]: is there a reason why you change the length here? just curious

File: pkg/pod/pod.go [hunk 3]
```diff
@@ -322,6 +327,36 @@ func (b *Builder) Build(ctx context.Context, taskRun *v1beta1.TaskRun, taskSpec
 		return nil, err
 	}
 
+	readonly := true
+	if config.FromContextOrDefaults(ctx).FeatureFlags.EnforceNonfalsifiability == config.EnforceNonfalsifiabilityWithSpire {
```
Comment (jerop) [inline pkg/pod/pod.go:331]: nit: given that you do this check multiple times, might be useful to define a func in feature_flags.go and reuse it, for example:

https://github.com/tektoncd/pipeline/blob/a28eff704fd3f58ffe5179ce6e96afa031b8b781/pkg/apis/config/feature_flags.go#L326-L338

Comment (jerop) [inline pkg/pod/pod.go:331]: nit: given that you do this check multiple times, might be useful to define a func in feature_flags.go and reuse it, for example:

https://github.com/tektoncd/pipeline/blob/a28eff704fd3f58ffe5179ce6e96afa031b8b781/pkg/apis/config/feature_flags.go#L326-L338

**Review by jerop — COMMENTED** [review]
@jagathprakash is this the last pull request implementing TEP-0089?

File: pkg/pod/pod.go (hunk 3 above)
Comment (jagathprakash) [reply pkg/pod/pod.go:331]: Done, defined a function IsSpireEnabled in feature_flags.go

File: examples/v1beta1/pipelineruns/4808-regression.yaml (hunk 2 above)
Comment (jagathprakash) [reply examples/v1beta1/pipelineruns/4808-regression.yaml:95]: The signature and the size of the certificate takes up approximately 1k. More details here https://github.com/tektoncd/community/blob/main/teps/0089-nonfalsifiable-provenance-support.md#signing-results.
Hence we reduce the number of bytes for the actual result.

File: cmd/imagedigestexporter/main.go (hunk 1 above)
Comment (jagathprakash) [reply cmd/imagedigestexporter/main.go:37]: The feature flag which is used in the config map is enforce-nonfalsifiability.
But this is a command line flag passed into imagedigestexporter by the controller. 
This command line flag is internal and not an API for the user, as such for these internal flags, enable_spire was retained as it keeps things simpler.

**Comment by jerop** [comment]
@jagathprakash fyi started taking a look last week but waiting for you to mark this as "ready for review" before taking another look

**Comment by jagathprakash** [comment]
> @jagathprakash fyi started taking a look last week but waiting for you to mark this as "ready for review" before taking another look

Thanks @jerop for taking a look. 
The PR is now ready to be reviewed.

File: test/taskrun_test.go [hunk 4]
```diff
@@ -210,3 +225,131 @@ spec:
...
+// the text modifies the taskrun results.
+// This change is caught by the taskrun reconciler when it tries to verify the results.
+func TestTaskRunModification(t *testing.T) {
+	ctx := context.Background()
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	var c *clients
+	var namespace string
+	var oldConfigMap map[string]string
```
Comment (chitrangpatel) [inline test/taskrun_test.go:240]: You might want to reset `results-from` to `sidecar-logs` in oldConfigMap since the default is `termination-message`

Comment (jagathprakash) [reply test/taskrun_test.go:240]: setupWithFlags will save the oldFlags and resetFlags will reset the old flags after execution.
If we set resuls-from to sidecar-logs in reset, we will incorrectly be setting so even in beta mode.

File: examples/v1beta1/pipelineruns/4808-regression.yaml (hunk 2 above)
Comment (chitrangpatel) [reply examples/v1beta1/pipelineruns/4808-regression.yaml:95]: With sidecar logs, we would not have to reduce this right? This is because of termination message extraction method?

File: pkg/pod/pod.go [hunk 5]
```diff
@@ -322,6 +327,36 @@ func (b *Builder) Build(ctx context.Context, taskRun *v1beta1.TaskRun, taskSpec
...
+		}
+		for i := range initContainers {
+			c := &initContainers[i]
+			c.VolumeMounts = append(c.VolumeMounts, corev1.VolumeMount{
+				Name:      spire.WorkloadAPI,
+				MountPath: spire.VolumeMountPath,
+				ReadOnly:  true,
+			})
+		}
+	}
```
Comment (chitrangpatel) [inline pkg/pod/pod.go:358]: Question: If users define sidecars, do we also need to mount these volumes there? Can users even produce results from sidecars (probably not so we may not want to worry)? Eventually, when we enable spire for sidecar logs, we will have to mount this volume in that specific sidecar.

File: pkg/pod/status.go [hunk 6]
```diff
@@ -109,11 +110,16 @@ func SidecarsReady(podStatus corev1.PodStatus) bool {
 }
 
 // MakeTaskRunStatus returns a TaskRunStatus based on the Pod's status.
-func MakeTaskRunStatus(ctx context.Context, logger *zap.SugaredLogger, tr v1beta1.TaskRun, pod *corev1.Pod, kubeclient kubernetes.Interface, ts *v1beta1.TaskSpec) (v1beta1.TaskRunStatus, error) {
+func MakeTaskRunStatus(ctx context.Context, logger *zap.SugaredLogger, tr v1beta1.TaskRun, pod *corev1.Pod,
+	kubeclient kubernetes.Interface, ts *v1beta1.TaskSpec, spireEnabled bool, spireAPI spire.ControllerAPIClient) (v1beta1.TaskRunStatus, error) {
```
Comment (chitrangpatel) [inline pkg/pod/status.go:114]: (nit) was this new line intentional?

**Review by chitrangpatel — APPROVED** [review]

File: pkg/pod/pod.go (hunk 5 above)
Comment (jagathprakash) [reply pkg/pod/pod.go:358]: Thats a good call out. Sidecars are given access to /tekton/results by the volume mounts.
Currently we are not mounting the SPIRE volumes into user defined sidecars as that would mean that the sidecars will have to initiate the signing, which might not work.
But for sidecar logs we will need to mount these volumes.

File: pkg/pod/status.go (hunk 6 above)
Comment (jagathprakash) [reply pkg/pod/status.go]: That was intentional as the line was too long. 
Go style guide suggests that long lines should just be retained as long lines.
https://google.github.io/styleguide/go/guide#line-length

Hence reverting back to long line here.

File: test/e2e-tests.sh [hunk 7]
```diff
@@ -53,6 +68,9 @@ function set_feature_gate() {
   fi
   printf "Setting feature gate to %s\n", ${gate}
   jsonpatch=$(printf "{\"data\": {\"enable-api-fields\": \"%s\"}}" $1)
+  if [ "$gate" == "alpha" ]; then
+    jsonpatch=$(printf "{\"data\": {\"enable-api-fields\": \"%s\"}}" $1)
+  fi
```
Comment (chitrangpatel) [inline test/e2e-tests.sh:73]: Are you doing anything new here? Doesn't seem like it. Unless I'm missing something obvious.

Comment (jagathprakash) [reply test/e2e-tests.sh:73]: Thats right. It was earlier enabling spire as well, but after discussions with @jerop, decided that we should only enable spire for spire tests.

File: examples/v1beta1/pipelineruns/4808-regression.yaml (hunk 2 above)
Comment (jagathprakash) [reply examples/v1beta1/pipelineruns/4808-regression.yaml:95]: I think we should have a test for sidecar logs which tests the maximum size of the result with sidecars.
With that we will still have to reduce the size of the result (i.e. maximum result) to accommodate the signatures of the results.

File: test/e2e-tests.sh [hunk 8]
```diff
@@ -40,6 +40,21 @@ header "Setting up environment"
...
+  if [ "$gate" == "alpha" ] ; then
+    DIR="$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )"
+    printf "Setting up environment for alpha features"
+    install_spire
+    patch_pipline_spire
+    kubectl apply -n tekton-pipelines -f "$DIR"/testdata/spire/config-spire.yaml
+    failed=0
+  fi
+}
```
Comment (jerop) [inline test/e2e-tests.sh:57]: did you mean for this function to be about enabling spire? 🤔

File: test/featureflags.go [hunk 9]
```diff
@@ -13,6 +13,39 @@ import (
 	"knative.dev/pkg/system"
 )
 
+var spireFeatureGates = map[string]string{
+	"enable-nonfalsifiability": "spire",
+	"enable-api-fields":        "alpha",
+}
```
Comment (jerop) [inline test/featureflags.go:19]: spire requires both of these feature flags to be enabled?

Comment (jerop) [reply test/featureflags.go:19]: what if I set `"enable-nonfalsifiability": "spire"`, but `"enable-api-fields": "stable"`?

File: docs/spire.md [hunk 10]
```diff
@@ -6,13 +6,58 @@ weight: 1660
 -->
 ⚠️ This is a work in progress: SPIRE support is not yet functional
 
-TaskRun result attestations is currently an alpha experimental feature. Currently all that is implemented is support for configuring Tekton to connect to SPIRE. See TEP-0089 for details on the overall design and feature set.
+TaskRun result attestations is currently an alpha experimental feature. Currently all that is implemented is support for configuring Tekton to connect to SPIRE and enabling TaskRun to sign and verify the TaskRunStatus. See TEP-0089 for details on the overall design and feature set.
 
 This being a large feature, this will be implemented in the following phases. This document will be updated as we implement new phases.
```
Comment (jerop) [inline docs/spire.md:11]: should this comment be updated now that we've finished the implementation

File: docs/spire.md [hunk 11]
```diff
@@ -6,13 +6,58 @@ weight: 1660
 -->
 ⚠️ This is a work in progress: SPIRE support is not yet functional
 
-TaskRun result attestations is currently an alpha experimental feature. Currently all that is implemented is support for configuring Tekton to connect to SPIRE. See TEP-0089 for details on the overall design and feature set.
+TaskRun result attestations is currently an alpha experimental feature. Currently all that is implemented is support for configuring Tekton to connect to SPIRE and enabling TaskRun to sign and verify the TaskRunStatus. See TEP-0089 for details on the overall design and feature set.
```
Comment (jerop) [inline docs/spire.md:9]: Add a link for users who may not know where to find TEPs:

```suggestion
TaskRun result attestations is currently an alpha experimental feature. Currently all that is implemented is support for configuring Tekton to connect to SPIRE and enabling TaskRun to sign and verify the TaskRunStatus. See [TEP-0089] for details on the overall design and feature set.

 [TEP-0089]: https://github.com/tektoncd/community/blob/main/teps/0089-nonfalsifiable-provenance-support.md
```

File: docs/spire.md [hunk 12]
```diff
@@ -6,13 +6,58 @@ weight: 1660
...
 
 This being a large feature, this will be implemented in the following phases. This document will be updated as we implement new phases.
 1.  Add a client for SPIRE (done).
 2.  Add a configMap which initializes SPIRE (in progress).
-3.  Modify TaskRun to sign and verify TaskRun Results using SPIRE.
-4.  Modify Tekton Chains to verify the TaskRun Results.
+3.  Modify TaskRun to sign and verify TaskRunStatus using SPIRE.
+4.  Modify Tekton Chains to verify the TaskRunStatus.
+
+When the TaskRun result attestations feature is enabled, all TaskRuns will produce a signature alongside its results, which can then be used to validate its provenance. For example, a TaskRun result that creates user-specified results `commit` and `url` would look like the following. `SVID`, `RESULT_MANIFEST`, `RESULT_MANIFEST.sig`, `commit.sig` and `url.sig` are generated attestations by the integration of SPIRE and Tekton Controller.
```
Comment (jerop) [inline docs/spire.md:17]: > TaskRun result attestations feature is enabled

which feature is this referring to? how can it be enabled? please be specific, and add a link to the feature

File: test/e2e-tests.sh (hunk 8 above)
Comment (jagathprakash) [reply test/e2e-tests.sh:57]: If we want to enable SPIRE we need to do a few changes to the pipelines deployment, in particular mount the spiffe-workload-api volume.
So for alpha tests this function makes all the changes to the deployment, which includes pipeline volume mounts and installing spire at the setup stage.
But it does not enable spire itself, i.e. it does not set the feature flag enforce_nonfalsifiability to spire, which ensures that all alpha tests except 2 in taskrun_test.go run without spire enabled.
For the 2 in taskrun_test.go we enable spire right in the test.

File: docs/spire.md (hunk 11 above)
Comment (jagathprakash) [reply docs/spire.md:9]: +1. 
Added a link to the TEP-0089 string below.

File: docs/spire.md (hunk 12 above)
Comment (jagathprakash) [reply docs/spire.md:17]: Added a link on "enabled" to point to how to enable this feature.

File: test/featureflags.go (hunk 9 above)
Comment (jagathprakash) [reply test/featureflags.go:19]: If enable-api-fields is not alpha, we get an error and the "enable-nonfalsifiability" will be set to the default i.e. no enforcement.
https://github.com/tektoncd/pipeline/blob/2a15755e9331156631aef8e382122b4bf685a131/pkg/apis/config/feature_flags.go#L221

File: test/e2e-tests.sh (hunk 8 above)
Comment (jerop) [reply test/e2e-tests.sh:57]: yes, the name is misleading -- `add_spire` is better

**Comment by jerop** [comment]
@jagathprakash please try to maintain the code coverage - it seems there are some spire-related codepaths in taskrun.go that are not tested

File: docs/spire.md (hunk 10 above)
Comment (jagathprakash) [reply docs/spire.md:11]: Agreed, also updated the features completed with "done" status.

File: test/e2e-tests.sh (hunk 8 above)
Comment (jagathprakash) [reply test/e2e-tests.sh:57]: Done.

**Comment by jagathprakash** [comment]
> @jagathprakash please try to maintain the code coverage - it seems there are some spire-related codepaths in taskrun.go that are not tested

@jerop Added a test which tests all new lines added for SPIRE. PTAL.

File: pkg/pod/pod.go [hunk 13]
```diff
@@ -322,6 +327,36 @@ func (b *Builder) Build(ctx context.Context, taskRun *v1beta1.TaskRun, taskSpec
...
+			})
+		}
+		for i := range initContainers {
+			c := &initContainers[i]
+			c.VolumeMounts = append(c.VolumeMounts, corev1.VolumeMount{
+				Name:      spire.WorkloadAPI,
+				MountPath: spire.VolumeMountPath,
+				ReadOnly:  true,
+			})
+		}
```
Comment (jerop) [inline pkg/pod/pod.go:357]:
```suggestion
		for i := range initContainers {
                        // mount SPIRE's CSI volume to each Init Container
			c := &initContainers[i]
			c.VolumeMounts = append(c.VolumeMounts, corev1.VolumeMount{
				Name:      spire.WorkloadAPI,
				MountPath: spire.VolumeMountPath,
				ReadOnly:  true,
			})
		}
```

File: pkg/pod/pod.go (hunk 5 above)
Comment (jerop) [reply pkg/pod/pod.go:358]: Please add this consideration to the issue I mentioned above for SPIRE + larger Results via Sidecar logs 🙏🏾

File: pkg/pod/status.go [hunk 14]
```diff
@@ -109,11 +110,15 @@ func SidecarsReady(podStatus corev1.PodStatus) bool {
...
-func MakeTaskRunStatus(ctx context.Context, logger *zap.SugaredLogger, tr v1beta1.TaskRun, pod *corev1.Pod, kubeclient kubernetes.Interface, ts *v1beta1.TaskSpec) (v1beta1.TaskRunStatus, error) {
+func MakeTaskRunStatus(ctx context.Context, logger *zap.SugaredLogger, tr v1beta1.TaskRun, pod *corev1.Pod, kubeclient kubernetes.Interface, ts *v1beta1.TaskSpec, spireEnabled bool, spireAPI spire.ControllerAPIClient) (v1beta1.TaskRunStatus, error) {
 	trs := &tr.Status
 	if trs.GetCondition(apis.ConditionSucceeded) == nil || trs.GetCondition(apis.ConditionSucceeded).Status == corev1.ConditionUnknown {
 		// If the taskRunStatus doesn't exist yet, it's because we just started running
 		markStatusRunning(trs, v1beta1.TaskRunReasonRunning.String(), "Not all Steps in the Task have finished executing")
+
+		if spireEnabled {
+			markStatusSignedResultsRunning(trs)
+		}
```
Comment (jerop) [inline pkg/pod/status.go:121]: why are we marking it as running without checking that the Condition.Status == corev1.ConditionUnknown?

Comment (jerop) [reply pkg/pod/status.go:121]: or is it assumed that as long as a TaskRun is running, then the signing of results is also running? if so, please add a code comment to clarify

File: pkg/reconciler/taskrun/taskrun.go [hunk 15]
```diff
@@ -134,6 +137,20 @@ func (c *Reconciler) ReconcileKind(ctx context.Context, tr *v1beta1.TaskRun) pkg
 		// on the event to perform user facing initialisations, such has reset a CI check status
 		afterCondition := tr.Status.GetCondition(apis.ConditionSucceeded)
 		events.Emit(ctx, nil, afterCondition, tr)
+	} else if spireEnabled {
```
Comment (jerop) [inline pkg/reconciler/taskrun/taskrun.go:140]: it's strange that the if-else block has conditions that seem unrelated:
- taskrun has started
- spire is enabled

what happens if both are true, taskrun has not started and spire is enabled, what happens? I'd expect the spire logic to be executed after starting the taskrun 🤔

Comment (jagathprakash) [reply pkg/reconciler/taskrun/taskrun.go:140]: Here we are checking if the previous status matches the signature. 
If the taskrun has not started then this is the first reconcile and we would not have a signature on the status. We need to verify this annotation only on the 2nd reconcile onwards.

Comment (jerop) [reply pkg/reconciler/taskrun/taskrun.go:140]: please add a comment with this explanation so it's clear when someone wants to make changes later

File: examples/v1beta1/pipelineruns/4808-regression.yaml (hunk 2 above)
Comment (jagathprakash) [reply examples/v1beta1/pipelineruns/4808-regression.yaml:95]: https://github.com/tektoncd/pipeline/issues/6363
Done.

File: pkg/pod/pod.go (hunk 5 above)
Comment (jagathprakash) [reply pkg/pod/pod.go:361]: Updated this in the issue.

File: pkg/pod/status.go (hunk 14 above)
Comment (jagathprakash) [reply pkg/pod/status.go:121]: We are checking this above. This is the same as markStatusRunning.

File: pkg/reconciler/taskrun/taskrun.go (hunk 15 above)
Comment (jagathprakash) [reply pkg/reconciler/taskrun/taskrun.go:140]: Done.

File: pkg/pod/pod.go [hunk 16]
```diff
@@ -322,6 +327,39 @@ func (b *Builder) Build(ctx context.Context, taskRun *v1beta1.TaskRun, taskSpec
 		return nil, err
 	}
 
+	readonly := true
```
Comment (Yongxuanzhang) [inline pkg/pod/pod.go:330]: curious why do we need this?

File: pkg/pod/status.go [hunk 17]
```diff
@@ -112,11 +113,15 @@ func SidecarsReady(podStatus corev1.PodStatus) bool {
 }
 
 // MakeTaskRunStatus returns a TaskRunStatus based on the Pod's status.
-func MakeTaskRunStatus(ctx context.Context, logger *zap.SugaredLogger, tr v1beta1.TaskRun, pod *corev1.Pod, kubeclient kubernetes.Interface, ts *v1beta1.TaskSpec) (v1beta1.TaskRunStatus, error) {
+func MakeTaskRunStatus(ctx context.Context, logger *zap.SugaredLogger, tr v1beta1.TaskRun, pod *corev1.Pod, kubeclient kubernetes.Interface, ts *v1beta1.TaskSpec, spireEnabled bool, spireAPI spire.ControllerAPIClient) (v1beta1.TaskRunStatus, error) {
```
Comment (Yongxuanzhang) [inline pkg/pod/status.go:116]: we can get the spireEnabled from ctx, do we need to pass it here?

**Review by Yongxuanzhang — COMMENTED** [review]
Sorry this PR is so big, 😅 
Only took a look at several files

File: docs/spire.md [hunk 18]
```diff
@@ -6,13 +6,58 @@ weight: 1660
 -->
 ⚠️ This is a work in progress: SPIRE support is not yet functional
 
-TaskRun result attestations is currently an alpha experimental feature. Currently all that is implemented is support for configuring Tekton to connect to SPIRE. See TEP-0089 for details on the overall design and feature set.
+TaskRun result attestations is currently an alpha experimental feature. Currently all that is implemented is support for configuring Tekton to connect to SPIRE and enabling TaskRun to sign and verify the TaskRunStatus. See [TEP-0089](https://github.com/tektoncd/community/blob/main/teps/0089-nonfalsifiable-provenance-support.md) for details on the overall design and feature set.
```
Comment (jerop) [inline docs/spire.md:9]:
```suggestion
TaskRun result attestations is currently an alpha experimental feature. Currently we support configuring Tekton to connect to SPIRE and enabling TaskRun to sign and verify the TaskRunStatus. The next and final step is to enable Chains to verify TaskRun Results. See [TEP-0089](https://github.com/tektoncd/community/blob/main/teps/0089-nonfalsifiable-provenance-support.md) for details on the overall design and feature set.
```

File: docs/spire.md [hunk 19]
```diff
@@ -127,3 +172,204 @@ To enable TaskRun attestations:
...
+- Verifying the result manifest
+  - Verify the content of `RESULT_MANIFEST` with the field `RESULT_MANIFEST.sig` with the SVID public key
+  - Verify that there is a corresponding field for all items listed in `RESULT_MANIFEST` (besides SVID and `*.sig` fields)
+- Verify individual result fields
+  - For each of the items in the results, verify its content against its associated `.sig` field
+
+
+# TaskRun Status attestations
+
+Each TaskRun status that is written by the tekton-pipelines-controller will be signed to ensure that there is no external
```
Comment (jerop) [inline docs/spire.md:280]:
```suggestion
Each TaskRun status that is written by the Tekton Pipelines controller will be signed to ensure that there is no external
```

File: docs/spire.md [hunk 20]
```diff
@@ -127,3 +172,204 @@ To enable TaskRun attestations:
...
+tampering of the TaskRun status. Upon each retrieval of the TaskRun, the tekton-pipelines-controller checks if the status is initialized,
```
Comment (jerop) [inline docs/spire.md:281]:
```suggestion
tampering of the TaskRun status. Upon each retrieval of the TaskRun, the Tekton Pipelines controller checks if the status is initialized,
```

File: docs/spire.md [hunk 21]
```diff
@@ -127,3 +172,204 @@ To enable TaskRun attestations:
...
+and that the signature validates the current status.
+The signature and SVID will be stored as annotations on the TaskRun Status field, and can be verified by a client.
+
+The verification is done on every consumption of the TaskRun except when the TaskRun is uninitialized. When uninitialized, the 
+tekton-pipelines-controller is not influenced by fields in the status and thus will not sign incorrect reflections of the TaskRun.
```
Comment (jerop) [inline docs/spire.md:286]:
```suggestion
Tekton Pipelines controller is not influenced by fields in the status and thus will not sign incorrect reflections of the TaskRun.
```

File: docs/spire.md [hunk 22]
```diff
@@ -127,3 +172,204 @@ To enable TaskRun attestations:
...
+
+The spec and TaskRun annotations/labels are not signed when there are valid interactions from other controllers or users (i.e. cancelling taskrun).
```
Comment (jerop) [inline docs/spire.md:288]:
```suggestion
The TaskRun spec and annotations/labels are not signed when there are valid interactions from other controllers or users (i.e. cancelling taskrun).
```

File: docs/spire.md [hunk 23]
```diff
@@ -127,3 +172,204 @@ To enable TaskRun attestations:
...
+  <TRUNCATED>
+```
+
+## How is the status being verified
+
+The signature are being verified by the Tekton controller, the process of verification is as follows:
+
+- Verify status-hash fields
+  - verify `tekton.dev/status-hash` content against its associated `tekton.dev/status-hash-sig` field. If status hash does 
+    not match invalidate the `tekton.dev/verified = no` annotation will be added
```
Comment (jerop) [inline docs/spire.md:371]: is the `TaskRunResultsVerified` also modified?

**Review by jerop — COMMENTED** [review]
leaving docs review comments for now, will continue with the rest of the pr shortly

**Review by jerop — COMMENTED** [review]
@jagathprakash please update the release notes to follow the template, otherwise it won't be picked up in the release pipeline

https://raw.githubusercontent.com/tektoncd/pipeline/main/.github/pull_request_template.md

File: pkg/pod/status.go (hunk 14 above)
Comment (jerop) [reply pkg/pod/status.go:124]: above, we're checking that apis.ConditionSucceeded.Status == corev1.ConditionUnknown -- what I'm wondering is how this is connected to setting v1beta1.TaskRunConditionResultsVerified.Status to corev1.ConditionUnknown

File: pkg/pod/pod.go (hunk 16 above)
Comment (jagathprakash) [reply pkg/pod/pod.go:330]: The ReadOnly in the VolumeSource below requires a pointer to a bool variable. Hence created this here.

**Comment by jerop** [comment]
@jagathprakash is splitting this up into smaller PRs so we can close this
